feature(Fee Center Management)

lomamech opened the pull request:
Implement Distribution Auxiliaries Fees Centers to the Main Centers
-Implement Component Account Reference Select
-Implement component Fee Center Select
-Implement component Percentage Input
-Implement component Multiple Services Select
- Add function Round Decimal in Util Service
- Display of the messages of the different invalidation messages in the
  different cases the distribution
- Improved expense center report, added optional option to include manual
  distributions,
-Implement distribution key integration test
- Update and Refactor some E2E test
- Restaure Module Fee Center Management

closes #3138

lomamech committed:
feature(Fee Center Management)
- Implement Component Account Reference Select

lomamech committed:
feature(Fee Center Management) Implement component Fee Center Select

lomamech committed:
feature(Fee Center) Implement component Percentage Input

lomamech committed:
feature(Fee Center) Implement component Multiple Services Select

lomamech committed:
feature(Fee Center)
- Add function Round Decimal in Util Service

lomamech committed:
Improve Reports off Fee Center

lomamech committed:
feature(Fee Center Report)
- Improved expense center report, added optional option to include manual
  distributions,

lomamech committed:
Restaure env.development

lomamech committed:
Resolve Conflict

lomamech committed:
test(Distribution Key)
- Implement distribution key integration test

lomamech committed:
test(Fee Center Distribution)
- Set Data in Database for Test

lomamech committed:
Rebase with Master Step 3

lomamech committed:
Restaure fee Center API

lomamech committed:
test(Fee Center Distribution)
- Implement Distribution Fees Centers Management End 2 End Test
- Improve and fix Fees Centers E2E test

lomamech committed:
refactor(Fee Center Report)
- Update and Improve Fee Center Report

lomamech committed:
refactor(Fee Center)
- Update displays value for Profits Centers values, displays values
  instead creditor values

lomamech committed:
Fix Sql Query

lomamech committed:
Rebase with Master

lomamech committed:
Resolve conflict

lomamech committed:
Resolve conflict

lomamech committed:
Fix Error

lomamech committed:
Resolve Conflict

lomamech committed:
Resolve conflict with routes.js

lomamech committed:
Resolve Conflict

lomamech committed:
Rebase branch Master

lomamech committed:
Rebase with Branch Master

lomamech committed:
Rebase with Branch Master

lomamech committed:
Complete Review and resolving conflict

jniles commented on client/src/js/components/bhFeeCenterSelect.js:
```diff
@@ -0,0 +1,53 @@
+        $ctrl.feeCenters = feeCenters;
+
+        if ($ctrl.filter) {
+          if ($ctrl.principal) {
```
You can simplify this logic by moving the `$ctrl.prinicpal` check into the filter function.  Like this:
```js
if ($ctrl.filter) {
  $ctrl.feeCenters = feeCenters.filter(center => {
     return $ctrl.principal ? item.is_principal : !item.is_principal;
  });
}
```

jniles commented on client/src/js/components/bhFeeCenterSelect.js:
```diff
@@ -0,0 +1,53 @@
+      })
+      .catch(Notify.handleError)
+      .finally(() => {
+        $ctrl.loading = false;
```
Does anything actually use the `loading` property?

jniles commented on client/src/i18n/fr/form.json:
```diff
@@ -8,7 +8,10 @@
             "ADD_PAYROLL_CONFIGURATION": "Configurer une période des paiements",
             "ADD_RUBRIC_CONFIGURATION" : "Ajouter une configuration des rubriques",
             "ADD_WEEKEND_CONFIGURATION": "Ajouter une configuration des weekend",
+            "AUTOMATIC_BREACKDOWN": "Ventilation automatique des factures",
```
`BREACKDOWN` should be spelled `BREAKDOWN` 👍

jniles commented on client/src/i18n/fr/form.json:
```diff
@@ -497,7 +515,7 @@
             "PERCENT": "Pourcent",
             "PERIODES": "Périodes",
-            "PERIOD": "Périod",
+            "PERIOD": "Période",
```
👍

jniles commented on client/src/js/components/bhPercentageInput.js:
```diff
@@ -0,0 +1,31 @@
+  const isDefined = angular.isDefined;
+
+  // translated label for the form input
+  $ctrl.label = $ctrl.label || 'FORM.LABELS.VALUE';
```
This binding should be in the `$ctrl.onInit()` lifecycle method.

jniles commented on client/src/modules/templates/bhPercentageInput.tmpl.html:
```diff
@@ -0,0 +1,42 @@
+        type="number"
+        name="percentageInputValue"
+        ng-model="$ctrl.model"
+        ng-min="$ctrl.min || 0"
```
It would be better to put these in the `$ctrl.onInit`.  Like this:
```js
$ctrl.$onInit = () => {
  $ctrl.symbol = $ctrl.symbol || '%';
  $ctrl.min = $ctrl.min || 0;
  // etc...
};
```
That way, this is only defined a single time.

jniles commented on client/src/modules/templates/bhPercentageInput.tmpl.html:
```diff
@@ -0,0 +1,42 @@
+<ng-form name="bhPercentageInputForm">
+  <div class="form-group"
+    ng-class="{ 'has-error' : $ctrl.validationTrigger && bhPercentageInputForm.percentageInputValue.$invalid }" bh-percentage-input>
```
Instead of using `validationTrigger`, you should use `bhPercentageInputForm.$submitted`.

jniles commented on client/src/modules/templates/bhServicesMultipleSelect.tmpl.html:
```diff
@@ -0,0 +1,29 @@
+<div ng-form="ServicesMultipleForm" bh-services-multiple-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : $ctrl.validationTrigger && ServicesMultipleForm.services.$invalid }">
```
Instead of using `validationTrigger`, you should use `ServicesMultipleForm.$submitted`.

lomamech committed:
improvement(Fee Center Distribution)
- Review and Sanitaze Code with Climat Code Rules
`#

lomamech committed:
improvement(Fee Center)
- Improve and Sanitaze Code

lomamech committed:
Improve and Sanitaze Code

lomamech committed:
Sanitaze Code Climat codes rules

lomamech committed:
Sanitaze Code

lomamech committed:
Sanitaze and Clear Code

lomamech committed:
Sanitaze and Clear Code

lomamech committed:
refactor(Update Fee Center): Refactor and Improve Update Distributed Fee Center Management

lomamech committed:
Sanitaze Code

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -8,7 +8,10 @@
             "ADD_PAYROLL_CONFIGURATION": "Set up a payment period",
             "ADD_RUBRIC_CONFIGURATION" : "Add a Rubric configuration",
             "ADD_WEEKEND_CONFIGURATION": "Add Weekend Configuration",
+            "AUTOMATIC_BREACKDOWN": "Automatic Breakdown of invoices",
```
All these "BREAKDOWN"s should be spelled "BREAKDOWN".

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -8,7 +8,10 @@
+            "AUTOMATIC_BREACKDOWN": "Automatic Breakdown of invoices",
             "BACK": "Back",
+            "BREAKDOWN_AUTOMATIQUE": "Breakdown Automatique",
+            "BREAKDOWNS_PERCENTAGES": "Breakdowns in Percentages",
```
A better way to say this is "Breakdown by percentage". 👍

jniles commented on client/src/modules/templates/bhPercentageInput.tmpl.html:
```diff
@@ -0,0 +1,42 @@
+    <p class="help-block" ng-messages="bhPercentageInputForm.percentageInputValue.$error" ng-show="$ctrl.validationTrigger">
+
+      <!-- displays: "This value is required" -->
+      <span ng-message="required">
```
It looks like all these errors are already [messages.tmpl.html](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/templates/messages.tmpl.html).

Could we just use our regular validation for this?  Like:
```html
<div class="help-block" ng-messages="CreateForm.name.$error" ng-show="CreateForm.$submitted">                                                                                                    
  <div ng-messages-include="modules/templates/messages.tmpl.html"></div>                                                                                                                         
</div>
```

jniles commented on server/controllers/finance/distributionFeeCenter/automatic.js:
```diff
@@ -0,0 +1,74 @@
+  const transUuids = data.map(item => db.bid(item.uuid));
+
+  const sql = `
+    SELECT BUID(gl.uuid) AS trans_uuid, gl.trans_id, gl.debit_equiv, gl.credit_equiv,
```
Hmm... `trans_uuid` is a confusing name because transactions don't have a `uuid`.  They have an `id`.  The `uuid` is for each row in the general ledger/posting journal.  Could we name this field something different?  Why not just `uuid`?  Or `row_uuid`?

jniles commented on server/controllers/finance/distributionFeeCenter/breackDown.js:
```diff
@@ -0,0 +1,69 @@
+    });
+  });
+
+  const sql = `INSERT INTO fee_center_distribution (
```
Is `fee_center_distribution` going to be the same size of the `general_ledger`?  It seems like we are repeating the following information between the two tables:
 1. account_id
 2. debit_equiv
 3. credit_equiv
 3. trans_uuid -> general_ledger.uuid
 4. trans_id
 5. currency_id

Is there any way we can reduce the number of repeated columns here?

lomamech replied:
If you are in an institution that does not use the distribution keys in this case, the distribution_fee_center will have a lot more lines than the general_ledger

Each of this column is needed, 
![image](https://user-images.githubusercontent.com/8035588/48550190-531d3780-e8d2-11e8-9149-c9dc5cbf65c7.png)

In this illustration, we see that a single line of the general ledger distributed to three main centers gives 3 lines

jniles replied:
Okay... we'll see what Larry thinks.  Can we at least remove the currency_id?  If we are using the `debit_equiv` and `credit_equiv`, it should always be the enterprise currency, correct?

jniles commented on server/controllers/finance/distributionFeeCenter/getDistributed.js:
```diff
@@ -0,0 +1,55 @@
+function getDistributed(req, res, next) {
+  const options = req.query;
+  options.is_cost = options.typeFeeCenter;
+  options.limit = 100;
```
Why is this `limit` forced to be 100?

lomamech replied:
Since this table may have multiple lines, at the risk of having overflow, I have limited the row numbers to 100 if no filter is enabled

jniles commented on server/controllers/finance/reports/fee_center/index.js:
```diff
@@ -0,0 +1,127 @@
+// default report parameters
+const DEFAULT_PARAMS = {
+  csvKey : 'account_reference',
+  filename : 'TREE.ACCOUNT_REFERENCE_REPORT',
```
I think this renders the `FEE_CENTER_REPORT`, right?  You should change all references of `ACCOUNT_REFERENCE_REPORT` to `FEE_CENTER_REPORT`, if so.

jniles commented on server/controllers/finance/reports/fee_center/report.handlebars:
```diff
@@ -0,0 +1,515 @@
+            </tr>
+            {{#each principalProfitCenters}}
+              <tr>
+                <td class="text-right"> {{add @index 1}} </td>
```
Is this array sorted?  What determines the order of the `@index`?

Generally we try to avoid putting `@index` in the rendering since it can change based on the sort order.  If there is a real `id` would could use, that would be preferable.

lomamech replied:
It's just to display the number on a list

jniles commented on server/models/migrations/v0.8.0-v0.9.0/migrate.sql:
```diff
@@ -847,3 +840,107 @@ CREATE TABLE `tags`(
 
 INSERT INTO unit VALUES
 (217, 'Tags','TREE.TAGS','', 1,'/modules/tags/tags','/tags');
+=======
```
These `===` are because of a broken merge.  Go ahead and remove them.

jniles commented on server/models/schema.sql:
```diff
@@ -2081,12 +2081,91 @@ CREATE TABLE `department`(
+  `is_principal` tinyint(1) UNSIGNED DEFAULT 0,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `fee_center_1` (`label`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
We are using the charset `utf8mb4`.  Please use `utf8mb4` for all your tables.

jniles replied:
You should also change your server to use `utf8mb4` as well, to make sure your environment is the same as the ones in production.

jniles replied:
This still needs to be addressed.

jniles commented on client/src/js/components/bhPercentageInput.js:
```diff
@@ -0,0 +1,30 @@
+    controller : PercentageInputController,
+    bindings : {
+      currencyId : '<', // one-way binding
+      model : '=', // two way binding
```
In general, we try to avoid two way bindings since they unpredictable.  The new angular best practice (of two years or so) is to use one-way bindings, with a callback when things change.  Like this:
```js
bindings : {
  model : '<',
  onChange : '&',
}
```

That way, you know exactly when the data is changed and who changed it.

You don't have to fix this in this PR, but in the future, please use one way bindings.  We can work on this together if you need help creating one.

jniles commented on client/src/js/components/bhServicesMultipleSelect.js:
```diff
@@ -0,0 +1,47 @@
+    $ctrl.label = $ctrl.label || 'FORM.LABELS.SERVICES';
+
+    // fired when the Services has been selected or removed from the list
+    $ctrl.onChange = $ctrl.onChange;
```
Please remove this line.  It doesn't do anything.  It's like saying:
```js
1 = 1;
```

jniles commented on client/src/modules/distribution_center/modals/distribution.modal.js:
```diff
@@ -0,0 +1,99 @@
+
+  FeeCenters.read()
+    .then((feeCenter) => {
+      vm.principalFeeCenter = feeCenter.filter(item => {
```
Could you make this plural since it is an array?  `vm.principalFeeCenters`.

jniles commented on client/src/modules/distribution_center/modals/distribution.modal.js:
```diff
@@ -0,0 +1,99 @@
+      }
+
+      if (!vm.principalFeeCenter.length) {
+        vm.noPrincilFeeCenter = true;
```
This variable should be `vm.noPrincipalFeeCenters`.

jniles replied:
Actually... I can't figure out where this is used.  Do you ever use it?  If not, let's remove it.

lomamech replied:
It's Used in distribution.modal.html

jniles commented on client/src/modules/distribution_center/modals/distribution.modal.js:
```diff
@@ -0,0 +1,99 @@
+
+    vm.invalidDistribution = sumDistributed !== amountEquiv;
+
+    vm.diffAmount = (diffAmount > 0)
```
Isn't this the same as `vm.diffAmount = Math.abs(diffAmount)`?  If so, I think that `Math.abs()` is clearer about what you are doing.

jniles commented on client/src/modules/distribution_center/modals/distribution_key.modal.html:
```diff
@@ -0,0 +1,45 @@
+    <div ng-repeat="principal in DistributionKeyModalCtrl.principalFeeCenter" class="form-group">
+      <bh-percentage-input
+        label="{{principal.label}}"
+        symbol="%"
```
Won't percentage inputs always have `%` as a symbol?  If so, can we just move `%` into the component and not have that binding?

jniles commented on client/src/modules/distribution_center/distribution_center.service.js:
```diff
@@ -0,0 +1,176 @@
+      backdrop : 'static',
+      controller : 'DistributionKeyModalController as DistributionKeyModalCtrl',
+      resolve : {
+        seetings : () => data,
```
Could you spell this as `settings`?  `seetings` -> `settings`

jniles commented on client/src/modules/distribution_center/distribution_center.service.js:
```diff
@@ -0,0 +1,176 @@
+    return Modal.open({
+      templateUrl : 'modules/distribution_center/modals/setting_distribution.modal.html',
+      size : 'md',
+      animation : false,
```
The properties `size`, `keyboard`, and `backdrop` are all set to the [default values](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/js/app.js#L242).  You can go ahead and remove them.

jniles commented on test/data.sql:
```diff
@@ -965,3 +965,74 @@ INSERT INTO `user_role`(uuid, user_id, role_uuid)
+INSERT INTO `reference_fee_center` (`id`, `fee_center_id`, `account_reference_id`, `is_cost`) VALUES (22, 4, 9, 1);
+
+-- GENERAL LEDGER FOR DISTRIBUTION FEE CENTER TEST
+INSERT INTO `general_ledger` (`uuid`, `project_id`, `fiscal_year_id`, `period_id`, `trans_id`, `trans_id_reference_number`, `trans_date`, `record_uuid`, `description`, `account_id`, `debit`, `credit`, `debit_equiv`, `credit_equiv`, `currency_id`, `entity_uuid`, `reference_uuid`, `comment`, `transaction_type_id`, `user_id`, `cc_id`, `pc_id`, `created_at`, `updated_at`) VALUES (0xE70109CCE0DC11E89F ...
```
Are there any vouchers/invoices/cash payments for these records?  In BHIMA it isn't possible to have records in the posting journal without making invoices, cash payments, or invoices. 

It would be better to create records for these using the use `PostVoucher()`, `PostInvoice`, etc...

lomamech replied:
Yes, I have set Its, 
````
-- INVOICE FOR DISTRIBUTION FEE CENTER
INSERT INTO `invoice` (`project_id`, `uuid`, `cost`, `debtor_uuid`, `service_id`, `user_id`, `date`, `description`, `reversed`, `edited`, `created_at`) 
VALUES (1, 0x79B0393553C54498A5ECA8CA6DFEA7AC, 256.6200, 0x3BE232F9A4B94AF6984C5D3F87D5C107, 3, 1, '2018-11-05 10:26:05', 'Facture de Test 2 Patient (PA.TPA.2) pour 2 items dans le service Medecine Interne. ', 0, 0, '2018-11-05 10:26:32');
INSERT INTO `invoice_item` (`invoice_uuid`, `uuid`, `inventory_uuid`, `quantity`, `inventory_price`, `transaction_price`, `debit`, `credit`) 
VALUES (0x79B0393553C54498A5ECA8CA6DFEA7AC, 0x3B456F6FAA594CE5BD80C91647EB6CF9, 0xDCFBCD6DB50F48B385614A0B8A4EFBA5, 100, 1.9000, 1.9000, 0.0000, 190.0000);
INSERT INTO `invoice_item` (`invoice_uuid`, `uuid`, `inventory_uuid`, `quantity`, `inventory_price`, `transaction_price`, `debit`, `credit`) 
VALUES (0x79B0393553C54498A5ECA8CA6DFEA7AC, 0xC150EFE416144C428643BFEA00608800, 0xC560E79CE57A4BC9BA53406AA861F858, 40, 5.1200, 5.1200, 0.0000, 204.8000);
````

jniles reviewed (commented):
@lomamech, I've completed an initial review.  This is a ton of code!

I've made some comments, let me know if you have any questions.  We should see if Larry can review this at some point.

lomamech committed:
improvement(Report Fee Center)
- Display undistributed Costs and Profits

lomamech committed:
bug(FeeCenter)
- Fix error on coding about closed of Modal when cancel button are clicked

lomamech committed:
Expected to return a value at the end of arrow function

lomamech committed:
test(Cash Registry)
- Update value on Cash Registry End to End Test

lomamech committed:
tests(Cash Registry)
- Reset value for Cash Registry Spec

lomamech committed:
Complete

lomamech committed:
Rebase with Branch Master

lomamech committed:
Use Math.abs for calculation

lomamech committed:
Clear and Sanitaze Code

lomamech committed:
test(Account Reference) reset value in end to end test

lomamech committed:
Resolve conflict

jniles commented on client/src/js/components/bhAccountReferenceSelect.js:
```diff
@@ -0,0 +1,64 @@
+    $ctrl.label = $ctrl.label || 'FORM.LABELS.REFERENCE';
+
+    // fired when an account reference has been selected
+    $ctrl.onSelectCallback = $ctrl.onSelectCallback || angular.noop;
```
It looks like this is required (`&`) - if so, there is no need to have the `onSelectCallback` default.

jniles commented on client/src/modules/distribution_center/distribution_center.html:
```diff
@@ -0,0 +1,95 @@
+      </div> 
+      
+      <div class="toolbar-item"> 
+        <bh-filter-toggle on-toggle="DistributionCenterCtrl.toggleFilter()">
```
Good use of `bhFilterToggle` 👍

jniles commented on client/src/modules/distribution_center/modals/setting_distribution.modal.js:
```diff
@@ -0,0 +1,98 @@
+  '$state', 'DistributionCenterService', 'filters', 'ModalService',
+  'NotifyService', 'bhConstants', 'Store', 'util', '$uibModalInstance', '$translate'];
+
+function SettingDistributionModalController($state, DistributionCenter, filters, ModalService, Notify,
```
It looks like `ModalService` is never used here.  Could you remove it?

jniles commented on client/src/modules/distribution_center/distribution_key/distribution_key.js:
```diff
@@ -0,0 +1,146 @@
+  .controller('DistributionKeyController', DistributionKeyController);
+
+DistributionKeyController.$inject = [
+  'DistributionCenterService', 'ModalService', 'NotifyService', 'uiGridConstants',
```
It looks like `ModalService` isn't used here - could you remove that?

jniles commented on client/src/modules/distribution_center/distribution_key/distribution_key.js:
```diff
@@ -0,0 +1,146 @@
+      headerCellFilter : 'translate',
+    }, {
+      field : 'rate',
+      type : 'number',
```
Nice! 👍

jniles commented on client/src/modules/distribution_center/distribution_center_update.service.js:
```diff
@@ -0,0 +1,49 @@
+ * @description
+ * Encapsulates common requests to the /distribution_fee_center/ URL.
+ */
+function DistributionCenterUpdateService(Api, Filters, AppCache) {
```
Nice service 👍 

It is well laid out and very easy to read.  Good code!

jniles commented on server/controllers/finance/distributionFeeCenter/getDistributionKey.js:
```diff
@@ -0,0 +1,42 @@
+    .catch(next);
+}
+
+function allDistibutionKey() {
```
Looks like there is a small typo here: `Distibution` -> `Distribution`

jniles commented on server/controllers/finance/distributionFeeCenter/proceed.js:
```diff
@@ -0,0 +1,71 @@
+
+function proceed(req, res, next) {
+  const { data } = req.body;
+  const isDebtor = !!data.debit_equiv;
```
Would this check also work: `data.debit_equiv > 0`?  It might be clearer - the debit value must be positive instead of a boolean.

jniles commented on server/controllers/finance/distributionFeeCenter/proceed.js:
```diff
@@ -0,0 +1,71 @@
+    const debitEquivDistributed = isDebtor ? dataValues[principalCenterId] : 0;
+    const creditEquivDistributed = isDebtor ? 0 : dataValues[principalCenterId];
+
+    if (debitEquivDistributed || creditEquivDistributed) {
```
It seems like this check will always be true.  Will it ever be false?  What happens if we remove it?

jniles commented on server/controllers/finance/distributionFeeCenter/proceed.js:
```diff
@@ -0,0 +1,71 @@
+
+  const transaction = db.transaction();
+
+  if (dataToDistribute.length) {
```
If `dataToDistribute.length` is 0, should we be throwing an error?  It looks like we will call `transaction.execute()` without adding any queries in that case.

jniles commented on server/controllers/finance/distributionFeeCenter/referenceAccount.js:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  // fee_center_id
+  const filter = params.fee_center_id ? ` AND fee_center.id = ${params.fee_center_id} ` : '';
```
It looks like this is vulnerable to an SQL injection attack.  In theory, we could set `fee_center_id` to `1 OR 1=1` and get everything, right?

The best course of action would be to use `FilterParser`.  If that is too complicated in this case, you should instead use `db.escape(params.fee_center_id)` ([docs](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/lib/db/index.js#L232)) and it should be fine.  Like this:
```js
const filter = params.fee_center_id ? ` AND fee_center.id = ${db.escape(params.fee_center_id)} ` : '';

jniles commented on server/controllers/finance/distributionFeeCenter/configuration.js:
```diff
@@ -0,0 +1,67 @@
+    })
+    .then((rows) => {
+
+      if (rows.length) {
```
I don't think there is a need for this check.  If `rows.length` is 0, `rows.forEach()` will just skip it anyway.  So you can probably just remove the `if (rows.length) {` check.

jniles commented on server/controllers/finance/distributionFeeCenter/referenceAccount.js:
```diff
@@ -0,0 +1,115 @@
+      FROM account
+      WHERE account.number LIKE '${item.number}%' AND account.type_id IN (4,5))`);
+      });
+      const sqlGetAllAccountsValids = tabAccountsValids.join(' UNION ');
```
This seems really inefficient.  Normally, we get accounts using a `WHERE` query.  Is there anyway we can avoid using a `UNION` here?

lomamech replied:
@jniles This section allows to generate a dynamic query, sqlGetAllAccountsValids will generate a request of this kind 
![image](https://user-images.githubusercontent.com/8035588/48707588-364d7080-ec00-11e8-94bc-f15f291aae76.png)

jniles commented on server/controllers/finance/distributionFeeCenter/referenceAccount.js:
```diff
@@ -0,0 +1,115 @@
+        ${item.account_reference_id} AS account_reference_id, account.id AS account_id,
+        account.number, ${item.is_exception} AS is_exception, account.type_id
+        FROM account
+        WHERE account.number LIKE '${item.number}%' AND account.type_id IN (4,5))`);
```
Can we use account ids instead of `LIKE`?  The `LIKE` statement will perform a regex on the account number, which is less efficient than just finding by ID.  Why don't we know the account IDs here?

jniles commented on server/controllers/finance/distributionFeeCenter/setting.js:
```diff
@@ -0,0 +1,54 @@
+  data.user_id = req.session.user.id;
+  const distributionKey = [];
+
+  Object.keys(dataValues).forEach((principalCenterId) => {
```
Just FYI, you have a couple of different tools you can use to loop through objects besides `Object.keys(o).forEach(fn)`.  Here are a couple of them:

```js
// imagine we have this object
const o = { id : 1, name : 'jniles', rate : 1.15, ill : true };

// we can do this to loop through only keys
Object.keys(o).forEach(key => console.log(key));
// => prints 'id', 'name', 'rate', 'ill'

// or we can loop through only values:
Object.values(o).forEach(value => console.log(values));
// => prints '1', 'jniles', '1.15', 'true'

// or we can loop through both keys and values using entries()
Object.entries(o).forEach(([key, value]) => console.log(key, value));
// => prints 'id' 1, 'name' 'jniles', 'rate' '1.15', 'ill' true'
```

jniles replied:
The documentation has more examples: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

lomamech committed:
test(Accounts References)
- Reset Accounts References E2E tests

jniles commented on server/controllers/finance/feeCenter.js:
```diff
@@ -0,0 +1,229 @@
+// Lists
+function list(req, res, next) {
+  const sql = `
+    SELECT f.id, f.label, f.is_principal, GROUP_CONCAT(' ', LOWER(ar.abbr)) AS abbrs, 
```
Wow, I had never heard of `GROUP_CONCAT` before!  Awesome - this is really good to know about.

lomamech replied:
@mbayopanda  use this sometimes

jniles reviewed (commented):
This is probably as far as I can go just looking at the code.  We'll need to get feedback from out client to understand how to improve this next, if it needs to be improved.

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -8,7 +8,10 @@
             "ADD_WEEKEND_CONFIGURATION": "Add Weekend Configuration",
+            "AUTOMATIC_BREAKDOWN": "Automatic Breakdown of invoices",
             "BACK": "Back",
+            "BREAKDOWN_AUTOMATIQUE": "Breakdown Automatique",
```
`AUTOMATIC_BREAKDOWN` and `BREAKDOWN_AUTOMATIQUE` are so close, since the first is related to invoices, could you rename the first to `AUTO_INVOICES_BREAKDOWN`, and the second to `AUTO_BREAKDOWN`

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -8,7 +8,10 @@
+            "AUTOMATIC_BREAKDOWN": "Automatic Breakdown of invoices",
             "BACK": "Back",
+            "BREAKDOWN_AUTOMATIQUE": "Breakdown Automatique",
+            "BREAKDOWNS_PERCENTAGES": "Breakdown by percentage",
```
`BREAKDOWNS_PERCENTAGES` is in plural but the definition doesn't contains plural elements.

If `BREAKDOWNS_PERCENTAGES` is not for plural elements, could you rename to a singular version some thing like `BREAKDOWN_PERCENTAGE`.

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -198,6 +212,7 @@
             "UPLOAD_PICTURE_FAILED": "The selected file is not an picture",
             "USER_CASHBOX": "These are the cashboxes that you have permission for manage",
             "USER_DEPOT": "These are the depots that you have permission for manage",
+            "VALIDATE_DISTRIBUTION": "Please click here to validate the distribution of all Ancillary Cost Centers to the Main Cost Centers according to the respective distribution keys",
```
I learned that `Ancillary` is similar to `Auxiliary`  👍

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -515,6 +533,7 @@
             "PRICE": "Price",
             "PRICE_LIST": "Price List",
             "PRINCIPAL": "Principal",
+            "PRINCIPAL_2": "Principal",
```
Why do you have `PRINCIPAL` and `PRINCIPAL_2` with the same definition or value ? Could you just use `PRINCIPAL`.

lomamech replied:
The difference is in the French translation, one is masculine and the other is feminine
````
"PRINCIPAL": "Principale",
"PRINCIPAL_2": "Principal",
````

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -799,15 +824,20 @@
-            "ATTENTION_WAITING_LIST": "Attention, only employees configured can be placed in the waiting list of payroll",            
+            "ATTENTION_WAITING_LIST": "Attention, only employees configured can be placed in the waiting list of payroll",
+            "ATTENTION_SELECTED": "Warning you have not selected any element",
+            "ATTENTION_WAITING_LIST": "Attention, only employees configured can be placed in the waiting list of payroll",
```
This entry `ATTENTION_WAITING_LIST` is duplicated in lines 827 and 829

mbayopanda commented on client/src/js/components/bhPercentageInput.js:
```diff
@@ -0,0 +1,36 @@
+    bindings : {
+      currencyId : '<', // one-way binding
+      model : '=', // two way binding
+      validationTrigger : '<', // one-way binding
```
Does the `validationTrigger` is necessary since it is not used ?

mbayopanda commented on client/src/js/components/bhServicesMultipleSelect.js:
```diff
@@ -0,0 +1,44 @@
+      servicesIds : '<?',
+      label : '@?',
+      required : '<?',
+      validationTrigger : '<?',
```
Could you remove the `validationTrigger` since it is not used

mbayopanda commented on client/src/modules/distribution_center/distribution_center.js:
```diff
@@ -0,0 +1,298 @@
+  function load(filters) {
+    // flush error and loading states
+    vm.hasError = false;
+    toggleLoadingIndicator();
```
In this line the vm.loading must be set to true, instead of just toggle it.

lomamech replied:
vm.loading is initially assigned by false at the beginning of the script
````
  vm.loading = false;
````
and it becomes true when calling this function

mbayopanda commented on client/src/modules/distribution_center/distribution_center.service.js:
```diff
@@ -0,0 +1,176 @@
+  service.openDistributionModal = openDistributionModal;
+  service.cacheFilters = cacheFilters;
+  service.proceedDistribution = proceedDistribution;
+  service.proceedBreackDownPercent = proceedBreackDownPercent;
```
If it is possible could you replace all Breack in names of variables or any other elements by Break

mbayopanda commented on client/src/modules/distribution_center/distribution_key/distribution_key.js:
```diff
@@ -0,0 +1,146 @@
+  function load() {
+    // flush error and loading states
+    vm.hasError = false;
+    toggleLoadingIndicator();
```
Could you set the vm.loading to true instead of calling the function.

mbayopanda commented on client/src/modules/distribution_center/modals/distribution_key.modal.js:
```diff
@@ -0,0 +1,77 @@
+  '$state', 'NotifyService', 'DistributionCenterService', 'PayrollConfigurationService',
+  'ExchangeRateService', 'SessionService', '$uibModalInstance', 'PatientInvoiceService',
+  'FeeCenterService', 'util', 'settings', '$translate',
+];
```
Could you remove : PayrollConfigurationService, ExchangeRateService, PatientInvoiceService, and util since they are not used

mbayopanda commented on client/src/modules/distribution_center/modals/distribution_key.modal.js:
```diff
@@ -0,0 +1,77 @@
+function DistributionKeyModalController(
+  $state, Notify, DistributionCenter, Configuration,
+  Exchange, Session, ModalInstance, Invoices, FeeCenters, util, settings, $translate,
+) {
```
These parameters must be removed : Configuration, Exchange, Invoices and util because they are not used.

mbayopanda commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -0,0 +1,150 @@
+
+FeeCenterModalController.$inject = [
+  '$state', 'FeeCenterService', 'ModalService', 'NotifyService', 'appcache',
+];
```
The `ModalService` is never used, could you remove it ?

mbayopanda commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -0,0 +1,150 @@
+    if (vm.auxiliaryCenter) {
+      vm.isCostCenter = 0;
+      vm.isProfitCenter = 0;
+    }
```
Instead of creating two `if` statement, you could just write one `if` statement with all the code inside : 
```js
if (value) {
  vm.hasProfitCenter = !value;
  vm.hasCostCenter = !value;
  vm.isCostCenter = 0;
  vm.isProfitCenter = 0;
}
```

mbayopanda commented on client/src/modules/templates/bhAccountReferenceSelect.tmpl.html:
```diff
@@ -0,0 +1,34 @@
+<div ng-form="{{ $ctrl.name }}" bh-account-reference-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : $ctrl.validationTrigger && AccountReferenceForm.account_reference_id.$invalid }">
```
Is it possible to remove the $ctrl.validationTrigger, by using the form name.

You can perform that by replacing the `ng-form="{{ $ctrl.name }}"` by something like `ng-form="AccountReferenceForm"`.

And replace all `$ctrl.validationTrigger` by `AccountReferenceForm.$submitted`

mbayopanda commented on client/src/modules/templates/bhAccountReferenceSelect.tmpl.html:
```diff
@@ -0,0 +1,34 @@
+    <ui-select
+      name="account_reference_id"
+      ng-model="$ctrl.accountReferenceId"
+      on-select="$ctrl.onSelect($item, $model)"
```
If $model is not used, you could remove it

mbayopanda commented on client/src/modules/templates/bhFeeCenterSelect.tmpl.html:
```diff
@@ -0,0 +1,28 @@
+<div ng-form="FeeCenterForm" bh-fee-center-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : $ctrl.validateTrigger && FeeCenterForm.fee_center_id.$invalid }">
```
Could you replace `$ctrl.validateTrigger` by `FeeCenterForm.$submitted`

mbayopanda commented on client/src/modules/templates/bhFeeCenterSelect.tmpl.html:
```diff
@@ -0,0 +1,28 @@
+      </ui-select-choices>
+    </ui-select>    
+
+    <div class="help-block" ng-messages="FeeCenterForm.fee_center_id.$error" ng-show="$ctrl.validateTrigger">
```
Could you replace `$ctrl.validateTrigger` by `FeeCenterForm.$submitted`

mbayopanda commented on client/src/modules/templates/bhFiscalPeriodSelect.tmpl.html:
```diff
@@ -25,7 +25,7 @@
       ng-model="$ctrl.selectedPeriodFrom"
       name="periodFrom"
-      ng-options="periodFrom as periodFrom.hrLabel for periodFrom in $ctrl.periods"
+      ng-options="periodFrom as periodFrom.hrLabel for periodFrom in $ctrl.periods track by periodFrom.id"
```
👍

mbayopanda commented on client/src/modules/templates/bhServicesMultipleSelect.tmpl.html:
```diff
@@ -0,0 +1,29 @@
+<div ng-form="ServicesMultipleForm" bh-services-multiple-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : ServicesMultipleForm.$submitted && ServicesMultipleForm.services.$invalid }">
```
👍

mbayopanda reviewed (commented):
I made some comments about codes for the client side

mbayopanda commented on server/controllers/finance/distributionFeeCenter/automatic.js:
```diff
@@ -0,0 +1,74 @@
+      rows.forEach((row) => {
+        data.forEach((item) => {
+          if (row.row_uuid === item.uuid) {
+            dataToDistribute.push([
```
Is it possible to just use the `rows` array like : 
```js
rows.forEach(row => {
  dataToDistribute.push([
    ...
  ]);
});
```
Because it seems like the `rows` array contains already data defined by this condition : 
```js
if (row.row_uuid === item.uuid)
```

lomamech replied:
After analyzing the codes, the first element ```` rows ```` Get the data from the request, linking trans_uuid to billing for related services

and for the automatic distribution there is place of combined some elements of ```` data```` and ```` rows```` like ````account_id, is_cost, fee_center_id````

mbayopanda commented on server/controllers/finance/distributionFeeCenter/configuration.js:
```diff
@@ -0,0 +1,67 @@
+            }
+          });
+
+          item.amount = item.credit ? item.credit : item.debit;
```
Does this scenario can happen, having item.credit with 0 as value, and item.debit = undefined ?

Assume that, we just define value for the profit (`item.credit`) and this value is `0`, the code above will affect automatically `undefined` (`item.debit`) to `item.amount`.

lomamech replied:
This scenario can only happen if in the table general ledger has a line with the field debit_equiv or credit_equiv with undefined values,

lomamech committed:
improvement(Fee Center Management Review)
- Remove not necessary property onSelectCallBack in
  bhAccountReferenceSelect
- Remove unused function
- Remove not necessary property currency_id in fee Center Distribution
- Improve and Sanitaze some Codes
- Set throw new BadRequest when data value is empty
- Use FilterParser in getting reference Account value by account
- Update format for table fee_center Update CHARACTER
- Update integration test for distribution test add test for Throwing Bad
  request if data to distribute is empty

lomamech committed:
improvement(Fee Center)
- Sanitaze Code
- Improve label of translate key
- Improve component structures and templates

lomamech committed:
improvement(Fee Center) Use breakDowm Instead breackDown

lomamech committed:
Resolve Conflict

lomamech committed:
Remove unused variable

lomamech committed:
Update migrate SQL

lomamech committed:
test(Fee Center Management)
- Reset Data and Value for Integration test for Account

lomamech committed:
refactor(Data for Test)
- Refactor Data for integration test

lomamech committed:
Refactor Data for Test Integration

lomamech committed:
Refactor Data

lomamech committed:
improvement(Table ENGINE)
- Improve table engine use utf8mb4 instead utf8

lomamech committed:
refactor(Schema.sql)
- Refactor Size for column type VARCHAR

lomamech committed:
Resolve conflict in Migrate

lomamech committed:
Fix and Resolve Conflict

lomamech committed:
feature(Fee Center Management)
- Added the option to reset auxiliaries centers distribution keys to the main
  centers

lomamech committed:
test(Distribution Key)
- Implement Integration and End to End test for reset distribution key
